### PR TITLE
Usage of byte array instead of pin/pass String - Android

### PR DESCRIPTION
**Pull request by adamblack**

### CODE REVIEW

**Comment by hvge** [comment]
Hi Adam, thank you for your pull request, but before we accept your change, we need you to file our CLA: https://forms.gle/r715RoVDoji4GD7K7

Without it, we are unable to accept any contributions from you. While you remain the author of the change, we need you to transfer the usage right to us so that we can freely work with the software (for example, change the license later to any license we need, or use the software in our commercial software without asking your permission, etc.).

I'll submit my review in a separate comment.

File: Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java [hunk 1]
```diff
@@ -90,13 +92,17 @@ private PasswordTester() {}
...
      * @return Strength of the password
      * @throws WrongPasswordException if provided password is {@code null} or cannot be converted
      *         to C-string.
      */
-    public PasswordStrength testPassword(@NonNull String password) throws WrongPasswordException {
-        switch (testPasswordJNI(password)) {
+    public PasswordStrength testPassword(@NonNull byte[] passwordBytes) throws WrongPasswordException {
+
+        int bufferLength = passwordBytes.length;
+        ByteBuffer buffer = getSharedBuffer(passwordBytes, bufferLength);
```
Comment (hvge) [inline Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java:103]: The `length` is a part of byte array, so there's no need to introduce a special length parameter here. Basically, by passing array itself we also pass the length of the array to the function.

File: Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java [hunk 2]
```diff
@@ -150,13 +158,30 @@ public PinTestResult testPin(@NonNull String pin) throws WrongPinException {
...
+     * @param byteArray Password or PIN as byte array
+     * @param arrayLength length of an array
+     * @return Shared byte buffer between Java/JNI
+     */
+    private ByteBuffer getSharedBuffer(byte[] byteArray, int arrayLength) {
+        ByteBuffer bb = ByteBuffer.allocateDirect(arrayLength);
+        bb.put(byteArray);
+        // This step might be too much but it is better than to just set this reference to null and
+        // hope for "quick" GC
+        Arrays.fill(byteArray, (byte) 0);
```
Comment (hvge) [inline Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java:178]: We don't own the byte array, so library should not modify its content. As a library developers we don't know what application's doing afterwards with its password stored in the array. I would keep it as it is and note this to the documentation that application should properly cleanup the password stored in the array.

Look at [Password.validatePasswordComplexity()](https://github.com/wultra/powerauth-mobile-sdk/blob/develop/proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/core/Password.java#L271). That's the actual point where PA-SDK meets with the password meter.

File: Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java [hunk 3]
```diff
@@ -150,13 +158,30 @@ public PinTestResult testPin(@NonNull String pin) throws WrongPinException {
...
     /**
-     * Defines constants returned from {@link #testPasswordJNI(String)} method.
+     * Converts user's byte input into ByteBuffer
+     *
+     * @param byteArray Password or PIN as byte array
+     * @param arrayLength length of an array
+     * @return Shared byte buffer between Java/JNI
+     */
+    private ByteBuffer getSharedBuffer(byte[] byteArray, int arrayLength) {
+        ByteBuffer bb = ByteBuffer.allocateDirect(arrayLength);
```
Comment (hvge) [inline Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java:174]: I think that usage of ByteBuffer here is too heavy. We don't transfer big memory regions, so there's no need to share exactly the same memory region between JVM and the native code.

I would keep `byte[]` as a type passed to JNI. What we need is to take a special care when JVM does the copy of the buffer into the native code space. The `GetByteArrayElements()` function has `is_copy` parameter that tells us that array was copied, so the temporary buffer has to be erased.

**Review by hvge — CHANGES_REQUESTED** [review]
Note for readers, this PR is attempt to solve #44 on Android platform.

Unfortunately, I cannot accept this change, because the library still has to provide the existing API that use strings at input. The ticket #44 is about just to enhance the existing API with a new low-level methods. We simply don't want to force other developers to use byte-array based password validation, because not everyone has such strict requirements as we have.

**Comment by adamblack** [comment]
@hvge Everything should be addressed. On the Android side, validation is handled by a shared C method that uses GetByteArrayElements.

I have two questions regarding the next steps since it would mean larger refactoring:

Android:
Testing of methods with the same signatures/functionality is painful without some interface or extra helper method or at least injecting raw string and byte array maps to tests. I guess it can be figured out later but it still breaks my heart. :]

iOS:
Usage of UnsafePointer works fine however PinTestResult requires original Pin (internally used to get length) which kinda kills the purpose of the UnsafePointer - to cast/obtain original string from memory. This needs to be addressed before merge. Solution might be additional method with [Uint8] signature instead of UnsafePointer however it is not so elegant.

File: Source/src_ios/PasswordTester.swift [hunk 4]
```diff
@@ -62,9 +62,9 @@ public class PasswordTester {
     ///
     /// - Parameter pin: PIN to evaluate..
     /// - Returns: Result of the testing
-    public func testPin(_ pin: String) -> PinTestResult {
+    public func testPin(_ passcodePtr: UnsafePointer<Int8>) -> PinTestResult {
```
Comment (hvge) [inline Source/src_ios/PasswordTester.swift:65]: We should also keep string version of API, just like on Android.

**Review by hvge — CHANGES_REQUESTED** [review]
This is much better.

File: Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java [hunk 5]
```diff
@@ -114,6 +116,35 @@ public PasswordStrength testPassword(@NonNull String password) throws WrongPassw
...
+                return PasswordStrength.MODERATE;
+            case PassResultCode.GOOD:
+                return PasswordStrength.GOOD;
+            case PassResultCode.STRONG:
+                return PasswordStrength.STRONG;
+            case PassResultCode.WRONG_INPUT:
+                throw new WrongPasswordException();
+            default:
+                throw new WrongPasswordException("Unknown result returned.");
+        }
```
[the remark this answers is not in the document]
Comment (adamblack) [reply Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java]: Agree. I was not sure how "deep" I can cut into the SDK code so I kept it separate but it should be ok in the latest commit. It also fixes our issue with testing.

File: Source/src_native/jni/PasswordTester.cpp [hunk 6]
```diff
@@ -82,4 +86,56 @@ JNIEXPORT jint JNICALL Java_com_wultra_android_passphrasemeter_PasswordTester_te
...
 }
 
+void test_value(JNIEnv *jenv, TestScenario test_scenario, jint& result, jbyteArray array) {
+    jsize arr_length = jenv->GetArrayLength(array);
+    jboolean isCopyFlag;
+    char* value_original = (char*) jenv->GetByteArrayElements(array, &isCopyFlag);
+
+    if (value_original != NULL) {
+        char* value_to_test = (char*) calloc(arr_length + 1, 1);
+        memcpy(value_to_test, value_original, arr_length);
```
[the remark this answers is not in the document]
Comment (adamblack) [reply Source/src_native/jni/PasswordTester.cpp:70]: In general I also prefer some helper functions like strndup, but in cases like this I really want to know what every line of code is doing (no implementation differences). Calloc is probably just an old habit of mine when working with chars from the past. :]

File: Source/src_ios/PasswordTester.swift (hunk 4 above)
Comment (adamblack) [reply Source/src_ios/PasswordTester.swift:65]: I might be missing the point but I believe there is an implicit conversion of string to UnsafePointer when C-side code requires `const char *str` which is also our case. Thanks to that, I thought changes in the API should not bother our users. I am ok with implementing both, just trying to understand the situation.

Comment (hvge) [reply Source/src_ios/PasswordTester.swift:65]: You're absolutely right. I completely forgot that this Swift's feature exist. This is then correct and we should only update the documentation to encourage app developers to pass the strings to the function. I'm pretty sure that this API will make nervous a lot of regular app developers :)

**Comment by adamblack** [comment]
@hvge I forgot to add similar handling of pin length to iOS side of PIN API as well. Should be fine in the newest commit.

File: Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java [hunk 7]
```diff
@@ -26,6 +26,9 @@
 
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
```
Comment (hvge) [inline Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/android/passphrasemeter/PasswordTester.java:31]: One last thing, please hit "optimize imports" on the whole project. There are few unused imports forgot by you and by us.

File: Source/src_ios/PasswordTester.swift [hunk 8]
```diff
@@ -97,9 +102,8 @@ public class PasswordTester {
     ///
     /// - Parameter password: Password to test
     /// - Returns: Strength of the password
-    public func testPassword(_ password: String) -> PasswordStrength {
-        guard password.count > 0 else { return .veryWeak }
-        return PasswordStrength(WPM_testPassword(password))
+    public func testPassword(_ passwordPtr: UnsafePointer<Int8>) -> PasswordStrength {
```
Comment (hvge) [inline Source/src_ios/PasswordTester.swift:105]: The same parameter naming inconsistency here.

**Review by hvge — CHANGES_REQUESTED** [review]
In general it looks OK now. I found just a small parameter naming inconsistency in Swift code.

Once you fix this I'll merge the PR

**Review by hvge — APPROVED** [review]

**Pull request merged**